src/SignalR/clients/ts/signalr/src/HttpConnection.ts

```diff
@@ -501,3 +511,96 @@ export class HttpConnection implements IConnection {
 function transportMatches(requestedTransport: HttpTransportType | undefined, actualTransport: HttpTransportType) {
     return !requestedTransport || ((actualTransport & requestedTransport) !== 0);
 }
+
+export class TransportSendQueue {
+    private buffer: any[] = [];
+    private sendBufferedData: PromiseSource;
+    private executing: boolean = true;
+    private transportResult: PromiseSource;
+
+    constructor(private transport: ITransport) {
+        this.sendBufferedData = new PromiseSource();
+        this.transportResult = new PromiseSource();
+
+        // Suppress typescript error about handling Promises. The sendLoop doesn't need to be observed
+        this.sendLoop().then(() => {}, (_) => {});
+    }
+
+    public send(data: string | ArrayBuffer): Promise<void> {
+        this.bufferData(data);
+        return this.transportResult.promise;
+    }
+
+    public stop(): void {
+        this.executing = false;
+        this.sendBufferedData.resolve();
+    }
+
+    private bufferData(data: string | ArrayBuffer): void {
+        if (this.buffer.length && typeof(this.buffer[0]) !== typeof(data)) {
+            throw new Error(`Expected data to be of type ${typeof(this.buffer)} but was of type ${typeof(data)}`);
+        }
+
+        this.buffer.push(data);
+        this.sendBufferedData.resolve();
+    }
+
+    private async sendLoop(): Promise<void> {
+        while (true) {
+            await this.sendBufferedData.promise;
+
+            if (!this.executing) {
+                // Did we stop?
+                break;
+            }
+
+            this.sendBufferedData = new PromiseSource();
+
+            const transportResult = this.transportResult;
+            this.transportResult = new PromiseSource();
+
+            const data = typeof(this.buffer[0]) === "string" ?
+                this.buffer.join("") :
+                TransportSendQueue.concatBuffers(this.buffer);
+
+            this.buffer.length = 0;
+
+            try {
+                await this.transport.send(data);
+                transportResult.resolve();
+            } catch (error) {
+                transportResult.reject(error);
+            }
+        }
+    }
+
+    private static concatBuffers(arrayBuffers: ArrayBuffer[]): ArrayBuffer {
+        const totalLength = arrayBuffers.map((b) => b.byteLength).reduce((a, b) => a + b);
+        const result = new Uint8Array(totalLength);
+        let offset = 0;
+        for (const item of arrayBuffers) {
+            result.set(new Uint8Array(item), offset);
+            offset += item.byteLength;
+        }
+
+        return result;
+    }
+}
+
+class PromiseSource {
```

Comment:
We already have this class somewhere

Reply:
It's in the test `Utiils`.  I do feel slightly icky about deriving from `Promise` when technically what we need is closer to `SemaphoreSlim.WaitAsync`